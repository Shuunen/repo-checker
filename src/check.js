import * as files from './files'
import { log } from './logger'
import { augmentData, getGitFolders } from './utils'

const Checkers = Object.keys(files).map(f => files[f])

function report (nbPassed, nbFailed) {
  log.info('Report :')
  log.setIndentLevel(1)
  log.test(nbPassed > 0, `${nbPassed} test(s) passed successfully`, false, true)
  log.test(nbFailed === 0, `${nbFailed} test(s) failed`, false, true)
  log.line()
  log.setIndentLevel(0)
  if (nbFailed > 0) throw new Error('failed at validating at least one rule in one folder')
}

export async function check (folderPath, data, doFix, doForce) {
  const folders = await getGitFolders(folderPath)
  let nbPassed = 0
  let nbFailed = 0
  for (const folder of folders) {
    log.info('Checking folder :', folder)
    log.setIndentLevel(1)
    const dataFolder = await augmentData(folder, data)
    for (const Checker of Checkers) {
      const instance = new Checker(folder, dataFolder, doFix, doForce)
      await instance.start()
      await instance.end()
      nbPassed += instance.nbPassed
      nbFailed += instance.nbFailed
    }
    log.line()
    log.setIndentLevel(0)
  }
  report(nbPassed, nbFailed)
}
