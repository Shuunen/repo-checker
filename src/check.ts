/* c8 ignore next */
import { ellipsis, green, red, yellow } from 'shuutils'
import type { ProjectData } from './constants'
import { DependencyCruiserFile, EditorConfigFile, EsLintFile, GitFile, GithubWorkflowFile, LicenseFile, NpmRcFile, NvmrcFile, NycRcFile, PackageJsonFile, ReadmeFile, RenovateFile, RepoCheckerConfigFile, TailwindFile, TravisFile, TsConfigFile } from './files/index.js'
import { log } from './logger'
import { augmentData, getProjectFolders } from './utils'

interface Indicators { failed: string[]; passed: string[]; warnings: string[] }

const checkers = [
  DependencyCruiserFile,
  EditorConfigFile,
  EsLintFile,
  GitFile,
  GithubWorkflowFile,
  LicenseFile,
  NpmRcFile,
  NvmrcFile,
  NycRcFile,
  PackageJsonFile,
  ReadmeFile,
  RenovateFile,
  RepoCheckerConfigFile,
  TailwindFile,
  TravisFile,
  TsConfigFile,
]

function reportLog (color: (string: string) => string, count: number, message: string): void {
  const line = `â€£ ${count} check${count > 1 ? 's' : ''} ${message}`
  /* c8 ignore next */
  log.info(count === 0 ? line : color(line))
}

interface CheckOptions {
  canFailStop?: boolean
  canFix?: boolean
  canForce?: boolean
  canThrow?: boolean
  data: ProjectData
  folderPath: string
}

function report ({ failed = [], passed = [], warnings = [] }: Indicators): void {
  log.info('Report :')
  reportLog(green, passed.length, 'are successful')
  reportLog(yellow, warnings.length, 'triggered warnings')
  reportLog(red, failed.length, 'are problematic')
  /* c8 ignore next */
}

// eslint-disable-next-line max-statements
export async function check ({ canFailStop = false, canFix = false, canForce = false, canThrow = true, data, folderPath }: CheckOptions): Promise<Indicators> {
  const folders = await getProjectFolders(folderPath)
  let passed: string[] = []
  let warnings: string[] = []
  let failed: string[] = []
  log.options.isActive = !data.isQuiet
  /* c8 ignore next */
  if (folders.length === 0) log.warn('no folder to check', folderPath)
  /* eslint-disable no-await-in-loop */
  for (const folder of folders) {
    if (canFailStop && failed.length > 0) {
      log.warn('stop checking other folders because of failures & --fail-stop option')
      break
    }
    log.info('Checking folder :', folder)
    const dataFolder = await augmentData(folder, data, folders.length > 1)
    for (const Checker of checkers) { // eslint-disable-line @typescript-eslint/naming-convention
      const instance = new Checker(folder, dataFolder, canFix, canForce)
      await instance.start()
      await instance.end()
      passed = [...passed, ...instance.passed]
      warnings = [...warnings, ...instance.warnings]
      failed = [...failed, ...instance.failed]
    }
  }
  /* eslint-enable no-await-in-loop */
  report({ failed, passed, warnings })
  const maxLogLength = 100
  if (canThrow && failed.length > 0) throw new Error(`failed at validating at least one rule in one folder : ${ellipsis(failed.join(', '), maxLogLength)}`)
  return { failed, passed, warnings }
}
