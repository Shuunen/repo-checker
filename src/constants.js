import { join } from 'path'

export const home = process.env.HOME
export const dataFileName = 'repo-checker.config.js'
export const repoCheckerPath = join(__dirname, '..')
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export const dataDefaults = {
  auto_merge: true,
  ban_sass: true,
  dev_deps_only: true,
  license: 'GPL-3.0',
  max_size_ko: 10,
  user_mail: 'romain.racamier@gmail.com',
  user_name: 'Romain Racamier-Lafon',
}
