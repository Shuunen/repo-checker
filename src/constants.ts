import { join } from 'path'

/* istanbul ignore next */
export const home = process.env.HOME ?? '' // when does HOME is not defined ?
export const dataFileName = 'repo-checker.config.js'
/* istanbul ignore next */
export const repoCheckerPath = process.env.pwd ?? process.cwd()
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export class ProjectData {
  auto_merge = true
  ban_sass = true
  dev_deps_only = true
  is_module = false
  license = 'GPL-3.0'
  max_size_ko = 1
  npm_package = false
  package_name = ''
  quiet = false
  repo_id = 'a-great-repo'
  use_typescript = false
  use_vue = false
  user_id = 'Shuunen'
  user_id_lowercase = 'shuunen'
  user_mail = 'romain.racamier@gmail.com'
  user_name = 'Romain Racamier-Lafon'
  web_published = false
  web_url = 'https://my-website.com'

  constructor (data: Partial<ProjectData> = {}) {
    Object.assign(this, data)
  }
}

export const dataDefaults = new ProjectData()
