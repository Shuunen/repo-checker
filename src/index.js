#!/usr/bin/env node

const arg = require('arg')
const path = require('path')

const check = require('./check')
const log = require('./logger')

function init () {
  const args = arg({ '--target': String, '--data': String, '--fix': Boolean }, { argv: process.argv.slice(2) })
  let target = args['--target']
  const doFix = args['--fix']
  if (!target) {
    log.info('\n', `no target specified via : --target=path/to/directory`)
    log.info(`targeting current directory...`, '\n')
    target = '.'
  }
  let data = {}
  if (args['--data']) {
    const p = path.join(__dirname, '..', args['--data'])
    log.info('loading data from', p)
    data = require(p)
  } else if (doFix) {
    log.warn('you should provide data to enhance fix')
  }
  log.start()
    .then(() => check(target, data, doFix))
    .catch(err => log.error(err))
    .then(() => log.end())
}

init()
