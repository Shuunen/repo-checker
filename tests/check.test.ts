import { test } from 'uvu'
import { equal } from 'uvu/assert'
import { check, report } from '../src/check'
import { ProjectData, repoCheckerPath } from '../src/constants'

test('repo-checker folder fails with low max size', async function () {
  const message = await check(repoCheckerPath, new ProjectData({ maxSizeKo: 2, npmPackage: true, quiet: true })).catch(() => 'failed')
  equal(message, 'failed')
})

test('repo-checker folder succeed with acceptable max size', async function () {
  const { passed, failed } = await check(repoCheckerPath, new ProjectData({ maxSizeKo: 120, npmPackage: true, quiet: true }))
  equal(passed, [
    'use-dependency-cruiser',
    'has-a-dependency-cruiser-js-file',
    'has-a-editorconfig-file',
    'editorconfig-has-space-indent',
    'editorconfig-has-indent-size-of-2',
    'editorconfig-has-unix-style-line-endings',
    'editorconfig-has-utf-8-encoding',
    'editorconfig-has-whitespace-trailing',
    'editorconfig-has-final-new-line-rule',
    'editorconfig-has-specific-markdown-trailing-rule',
    'editorconfig-has-no-specific-html-indent-rule',
    'has-no-xo-config-js-file',
    'eslintrc-json-has-eslint-recommended-rules-extend',
    'eslintrc-json-has-unicorn-rules-extend',
    'eslintrc-json-has-no-promise-plugin-require-eslint-7',
    'eslintrc-json-has-no-plugin-section-since-plugin-are-included-by-extends',
    'eslintrc-json-current-eslintrc-json-has-only-36-of-the-36-custom-rules-in-repo-checker-eslintrc-json',
    'has-a-gitignore-file',
    'gitignore-has-node-modules',
    'gitignore-has-no-pnpm-lock-exclusion',
    'github-workflows-ci-yml-has-a-checkout-step-in-ci-workflow',
    'github-workflows-ci-yml-has-a-node-step-in-ci-workflow',
    'github-workflows-ci-yml-has-a-pnpm-setup-step',
    'has-a-license-file',
    'license-has-a-gpl-title',
    'license-has-a-version-3-subtitle',
    'has-a-nvmrc-file',
    'nvmrc-has-a-recent-lts-node-version',
    'nycrc-file-exists',
    'has-a-nycrc-json-file',
    'nycrc-json-has-a-schema-declaration',
    'has-a-package-json-file',
    'package-json-main-file-maximum-size-is-specified-in-data-file-ex-max-size-ko-100',
    'package-json-has-a-dist-repo-check-min-cjs-file',
    'package-json-main-file-specified-in-package-json-dist-repo-check-min-cjs-exists-on-disk-be-sure-to-build-before-run-repo-check',
    'package-json-main-file-size-should-be-less-or-equal-to-max-size-allowed-120ko',
    'package-json-has-a-schema-declaration',
    'package-json-has-a-bugs-property',
    'package-json-has-a-description-property',
    'package-json-has-a-files-property',
    'package-json-has-a-homepage-property',
    'package-json-has-a-keywords-property',
    'package-json-has-a-private-property',
    'package-json-has-a-repository-property',
    'package-json-has-a-author-property',
    'package-json-has-a-name-property',
    'package-json-has-a-version-property',
    'package-json-has-a-license-property',
    'package-json-has-a-gpl-3-0-license',
    'package-json-has-no-engines-section',
    'package-json-has-a-script-section',
    'package-json-has-a-pre-script-for-version-automation',
    'package-json-has-a-post-script-for-version-automation',
    'package-json-has-a-typescript-build-or-check',
    'package-json-has-watchlist-eager-param',
    'package-json-has-a-ci-script',
    'package-json-has-no-sass-dependency-fat-useless',
    'package-json-has-no-cross-var-dependency-old-deprecated',
    'package-json-has-no-tslint-dependency-deprecated',
    'package-json-has-no-eslint-plugin-promise-5-dependency-require-eslint-7',
    'package-json-has-no-patch-precision',
    'package-json-one-unit-testing-dependency-from-mocha-uvu',
    'package-json-one-coverage-dependency-from-nyc-c8',
    'package-json-has-eslint-cache-flag',
    'package-json-assert-dependency-used-in-import-equal-from-uvu-assert-instead-works-also-as-deep-equal-alternative',
    'package-json-has-no-fat-color-dependency-use-shuutils-or-nanocolors',
    'package-json-has-no-fat-fs-extra-dependency-use-native-fs',
    'package-json-has-no-utopian-shuunen-stack-dependency',
    'package-json-has-no-fat-task-runner-use-npm-run-xyz-npm-run-abc-for-sequential-or-zero-deps-package-npm-parallel',
    'has-a-readme-md-file',
    'readme-md-has-a-title',
    'readme-md-has-no-link-to-deprecated-netlify-com',
    'readme-md-has-no-links-without-https-scheme',
    'readme-md-has-no-crlf-windows-carriage-return',
    'readme-md-has-no-star-flavored-list',
    'readme-md-has-a-project-license-badge',
    'readme-md-has-a-npm-monthly-downloads-badge',
    'readme-md-has-a-npm-version-badge',
    'readme-md-has-a-publish-size-badge',
    'readme-md-has-a-install-size-badge',
    'readme-md-has-a-thanks-section',
    'readme-md-has-a-thanks-to-shields-io',
    'readme-md-has-no-remaining-thanks-to-travis-ci-com',
    'readme-md-has-a-thanks-to-github',
    'readme-md-has-no-remaining-thanks-to-netlify',
    'readme-md-has-a-thanks-to-arg',
    'readme-md-has-no-remaining-thanks-to-ava',
    'readme-md-has-a-thanks-to-c8',
    'readme-md-has-no-remaining-thanks-to-chokidar',
    'readme-md-has-no-remaining-thanks-to-cypress-io',
    'readme-md-has-a-thanks-to-esbuild',
    'readme-md-has-a-thanks-to-eslint',
    'readme-md-has-no-remaining-thanks-to-mocha',
    'readme-md-has-no-remaining-thanks-to-npm-run-all',
    'readme-md-has-no-remaining-thanks-to-npm-parallel',
    'readme-md-has-no-remaining-thanks-to-nuxt',
    'readme-md-has-no-remaining-thanks-to-nyc',
    'readme-md-has-no-remaining-thanks-to-reef',
    'readme-md-has-a-thanks-to-repo-checker',
    'readme-md-has-no-remaining-thanks-to-rollup',
    'readme-md-has-no-remaining-thanks-to-servor',
    'readme-md-has-no-remaining-thanks-to-showdown',
    'readme-md-has-a-thanks-to-shuutils',
    'readme-md-has-no-remaining-thanks-to-tailwind-css',
    'readme-md-has-no-remaining-thanks-to-tsup',
    'readme-md-has-a-thanks-to-uv-u',
    'readme-md-has-no-remaining-thanks-to-vite',
    'readme-md-has-no-remaining-thanks-to-vue',
    'readme-md-has-a-thanks-to-watchlist',
    'readme-md-has-no-remaining-thanks-to-xo',
    'has-a-renovate-json-file',
    'renovate-json-has-a-schema-declaration',
    'renovate-json-has-an-extends-section',
    'renovate-json-has-a-base-config',
    'renovate-json-has-a-dashboard-setting-to-false',
    'renovate-json-has-an-auto-merge-preset',
    'renovate-json-has-a-preserve-semver-ranges-preset',
    'has-no-repo-checker-js-file',
    'has-a-repo-checker-config-js-file',
    'repo-checker-config-js-has-no-more-snake-case',
    'tsconfig-json-has-a-schema-declaration',
    'tsconfig-json-has-an-include-section',
    'tsconfig-json-has-a-allow-unreachable-code-compiler-option',
    'tsconfig-json-has-a-allow-unused-labels-compiler-option',
    'tsconfig-json-has-a-check-js-compiler-option',
    'tsconfig-json-has-a-es-module-interop-compiler-option',
    'tsconfig-json-has-a-exact-optional-property-types-compiler-option',
    'tsconfig-json-has-a-force-consistent-casing-in-file-names-compiler-option',
    'tsconfig-json-has-a-imports-not-used-as-values-compiler-option',
    'tsconfig-json-has-a-no-fallthrough-cases-in-switch-compiler-option',
    'tsconfig-json-has-a-no-implicit-any-compiler-option',
    'tsconfig-json-has-a-no-implicit-override-compiler-option',
    'tsconfig-json-has-a-no-implicit-returns-compiler-option',
    'tsconfig-json-has-a-no-property-access-from-index-signature-compiler-option',
    'tsconfig-json-has-a-no-unchecked-indexed-access-compiler-option',
    'tsconfig-json-has-a-no-unused-locals-compiler-option',
    'tsconfig-json-has-a-no-unused-parameters-compiler-option',
    'tsconfig-json-has-a-skip-lib-check-compiler-option',
    'tsconfig-json-has-a-strict-compiler-option',
    'tsconfig-json-has-a-out-dir-compiler-option',
    'tsconfig-json-has-a-module-resolution-compiler-option',
    'tsconfig-json-has-a-target-compiler-option',
    'tsconfig-json-has-a-non-empty-lib-compiler-option',
    'tsconfig-json-has-a-non-empty-types-compiler-option',
    'tsconfig-json-my-folder-is-not-needed-in-include-section-my-folder-is-enough',
  ], 'passed')
  equal(failed, [], 'failed')
})

test('report does nothing by default', function () {
  report()
})

test.run()
