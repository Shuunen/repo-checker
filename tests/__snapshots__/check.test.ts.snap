// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`check A folder fails with low max size 1`] = `"failed at validating at least one rule in one folder : package-json-main-file-size-45ko-should-be-less-or-equal-to-max-size-allowed-2ko"`;

exports[`check B folder succeed 1`] = `
[
  "use-dependency-cruiser",
  "has-a-dependency-cruiser-config-file",
  "has-a-editorconfig-file",
  "editorconfig-has-space-indent",
  "editorconfig-has-indent-size-of-2",
  "editorconfig-has-unix-style-line-endings",
  "editorconfig-has-utf-8-encoding",
  "editorconfig-has-whitespace-trailing",
  "editorconfig-has-final-new-line-rule",
  "editorconfig-has-no-specific-html-indent-rule",
  "has-no-xo-config-js-file",
  "eslintrc-json-has-hardcore-rules-extend",
  "eslintrc-json-has-unicorn-rules-extend",
  "eslintrc-json-has-no-promise-plugin-require-eslint-7",
  "eslintrc-json-current-eslintrc-json-has-only-77-of-the-77-custom-rules-in-repo-checker-eslintrc-json",
  "eslintrc-json-has-hardcore-typescript-rules-extend",
  "eslintrc-json-has-a-disabled-explicit-function-return-type",
  "avoid-main-branch-reference-use-master-instead-git-bclean",
  "has-a-gitignore-file",
  "gitignore-has-node-modules",
  "gitignore-has-no-pnpm-lock-exclusion",
  "github-workflows-ci-yml-has-a-checkout-step-in-ci-workflow",
  "github-workflows-ci-yml-has-a-node-step-in-ci-workflow",
  "github-workflows-ci-yml-has-a-pnpm-setup-step",
  "github-workflows-ci-yml-has-no-main-branch-reference",
  "has-a-license-file",
  "license-has-a-gpl-title",
  "license-has-a-version-3-subtitle",
  "npmrc-file-should-be-present-when-using-pre-post-task-in-package-json-to-set-enable-pre-post-scripts-option",
  "npmrc-has-enable-pre-post-scripts-option",
  "has-a-nvmrc-file",
  "nvmrc-has-a-recent-lts-node-version",
  "has-a-package-json-file",
  "package-json-main-file-maximum-size-is-specified-in-data-file-ex-max-size-ko-100",
  "package-json-has-a-dist-repo-check-min-cjs-file",
  "package-json-main-file-specified-in-package-json-dist-repo-check-min-cjs-exists-on-disk-be-sure-to-build-before-run-repo-check",
  "package-json-main-file-size-45ko-should-be-less-or-equal-to-max-size-allowed-120ko",
  "package-json-has-a-schema-declaration",
  "package-json-has-a-bugs-property",
  "package-json-has-a-description-property",
  "package-json-has-a-files-property",
  "package-json-has-a-homepage-property",
  "package-json-has-a-keywords-property",
  "package-json-has-a-private-property",
  "package-json-has-a-repository-property",
  "package-json-has-a-author-property",
  "package-json-has-a-name-property",
  "package-json-has-a-version-property",
  "package-json-has-a-license-property",
  "package-json-has-a-gpl-3-0-license",
  "package-json-has-a-typescript-build-or-check",
  "package-json-has-a-typescript-lint",
  "package-json-has-a-script-section",
  "package-json-has-a-pre-script-for-version-automation",
  "package-json-has-a-post-script-for-version-automation",
  "package-json-has-a-depcruise-usage",
  "package-json-has-a-unique-mark-task",
  "package-json-has-a-check-script",
  "package-json-has-a-final-echo-for-task-build",
  "package-json-has-a-final-echo-for-task-check",
  "package-json-has-a-final-echo-for-task-lint",
  "package-json-has-a-final-echo-for-task-test",
  "package-json-has-no-sass-dependency-fat-useless",
  "package-json-has-no-cross-var-dependency-old-deprecated",
  "package-json-has-no-tslint-dependency-deprecated",
  "package-json-has-no-eslint-plugin-promise-5-dependency-require-eslint-7",
  "package-json-has-no-patch-precision",
  "package-json-one-unit-testing-dependency-from-vitest-mocha-uvu",
  "package-json-one-coverage-dependency-from-nyc-c8",
  "package-json-has-eslint-cache-flag",
  "package-json-has-no-eslint-ignore-flag-solution-1-just-remove-it-useless-most-of-the-time-check-debugeslint-cli-engine-npx-eslint-to-see-linted-files-or-solution-2-use-ignore-patterns-inside-eslintrc-json-the-objective-here-is-to-let-the-eslint-cli-vscode-eslint-use-the-same-config",
  "package-json-has-no-fat-color-dependency-use-shuutils-or-nanocolors",
  "package-json-has-no-fat-fs-extra-dependency-use-native-fs",
  "package-json-has-no-utopian-shuunen-stack-dependency",
  "package-json-has-no-fat-slow-jsdom-dependency-use-happy-dom-instead",
  "package-json-has-no-fat-task-runner-use-npm-run-xyz-npm-run-abc-for-sequential-or-zero-deps-package-npm-parallel",
  "has-a-readme-md-file",
  "readme-md-has-a-title",
  "readme-md-has-no-link-to-deprecated-netlify-com",
  "readme-md-has-no-links-without-https-scheme",
  "readme-md-has-no-crlf-windows-carriage-return",
  "readme-md-has-no-star-flavored-list",
  "readme-md-has-a-project-license-badge",
  "readme-md-has-a-npm-monthly-downloads-badge",
  "readme-md-has-a-npm-version-badge",
  "readme-md-has-a-publish-size-badge",
  "readme-md-has-a-install-size-badge",
  "readme-md-has-a-thanks-section",
  "readme-md-has-a-thanks-to-shields-io",
  "readme-md-has-no-remaining-thanks-to-travis-ci-com",
  "readme-md-has-a-thanks-to-github",
  "readme-md-has-no-remaining-thanks-to-netlify",
  "readme-md-has-a-thanks-to-arg",
  "readme-md-has-no-remaining-thanks-to-ava",
  "readme-md-has-a-thanks-to-c8",
  "readme-md-has-no-remaining-thanks-to-chokidar",
  "readme-md-has-no-remaining-thanks-to-cypress-io",
  "readme-md-has-a-thanks-to-dependency-cruiser",
  "readme-md-has-a-thanks-to-esbuild",
  "readme-md-has-a-thanks-to-eslint",
  "readme-md-has-no-remaining-thanks-to-mocha",
  "readme-md-has-no-remaining-thanks-to-npm-run-all",
  "readme-md-has-no-remaining-thanks-to-npm-parallel",
  "readme-md-has-no-remaining-thanks-to-nuxt",
  "readme-md-has-no-remaining-thanks-to-nyc",
  "readme-md-has-no-remaining-thanks-to-reef",
  "readme-md-has-a-thanks-to-repo-checker",
  "readme-md-has-no-remaining-thanks-to-rollup",
  "readme-md-has-no-remaining-thanks-to-servor",
  "readme-md-has-no-remaining-thanks-to-showdown",
  "readme-md-has-a-thanks-to-shuutils",
  "readme-md-has-no-remaining-thanks-to-tailwind-css",
  "readme-md-has-no-remaining-thanks-to-tsup",
  "readme-md-has-no-remaining-thanks-to-uv-u",
  "readme-md-has-a-thanks-to-vitest",
  "readme-md-has-no-remaining-thanks-to-vite",
  "readme-md-has-no-remaining-thanks-to-vue",
  "readme-md-has-no-remaining-thanks-to-watchlist",
  "readme-md-has-no-remaining-thanks-to-xo",
  "has-a-renovate-json-file",
  "renovate-json-has-a-schema-declaration",
  "renovate-json-has-an-extends-section",
  "renovate-json-has-a-base-config",
  "renovate-json-has-a-dashboard-setting-to-false",
  "renovate-json-has-an-auto-merge-preset",
  "renovate-json-has-a-preserve-semver-ranges-preset",
  "has-a-repo-checker-json-file",
  "tsconfig-json-has-a-schema-declaration",
  "tsconfig-json-has-an-include-section",
  "tsconfig-json-has-a-allow-unreachable-code-compiler-option",
  "tsconfig-json-has-a-allow-unused-labels-compiler-option",
  "tsconfig-json-has-a-check-js-compiler-option",
  "tsconfig-json-has-a-es-module-interop-compiler-option",
  "tsconfig-json-has-a-exact-optional-property-types-compiler-option",
  "tsconfig-json-has-a-force-consistent-casing-in-file-names-compiler-option",
  "tsconfig-json-has-a-no-fallthrough-cases-in-switch-compiler-option",
  "tsconfig-json-has-a-no-implicit-any-compiler-option",
  "tsconfig-json-has-a-no-implicit-override-compiler-option",
  "tsconfig-json-has-a-no-implicit-returns-compiler-option",
  "tsconfig-json-has-a-no-property-access-from-index-signature-compiler-option",
  "tsconfig-json-has-a-no-unchecked-indexed-access-compiler-option",
  "tsconfig-json-has-a-no-unused-locals-compiler-option",
  "tsconfig-json-has-a-no-unused-parameters-compiler-option",
  "tsconfig-json-has-a-skip-lib-check-compiler-option",
  "tsconfig-json-has-a-strict-compiler-option",
  "tsconfig-json-has-a-out-dir-compiler-option",
  "tsconfig-json-has-a-module-resolution-compiler-option",
  "tsconfig-json-has-a-target-compiler-option",
  "tsconfig-json-has-a-non-empty-lib-compiler-option",
  "tsconfig-json-has-a-non-empty-types-compiler-option",
  "tsconfig-json-my-folder-is-not-needed-in-include-section-my-folder-is-enough",
]
`;

exports[`check B folder succeed 2`] = `[]`;

exports[`check B folder succeed 3`] = `
[
  "readme-md-could-have-a-svg-logo-logofolder-logo-svg",
  "readme-md-could-have-a-demo-screen-or-gif-demofolder-demo-gif",
]
`;

exports[`check C data/tsProject 1`] = `
[
  "editorconfig-has-no-specific-html-indent-rule",
  "has-no-xo-config-js-file",
  "eslintrc-json-has-hardcore-rules-extend",
  "eslintrc-json-has-no-promise-plugin-require-eslint-7",
  "eslintrc-json-has-hardcore-typescript-rules-extend",
  "eslintrc-json-has-a-disabled-explicit-function-return-type",
  "avoid-main-branch-reference-use-master-instead-git-bclean",
  "gitignore-has-no-pnpm-lock-exclusion",
  "has-a-package-json-file",
  "package-json-has-a-description-property",
  "package-json-has-a-author-property",
  "package-json-has-a-version-property",
  "package-json-has-a-license-property",
  "package-json-has-a-mit-license",
  "package-json-has-a-script-section",
  "package-json-has-no-sass-dependency-fat-useless",
  "package-json-has-no-cross-var-dependency-old-deprecated",
  "package-json-has-no-tslint-dependency-deprecated",
  "package-json-has-no-eslint-plugin-promise-5-dependency-require-eslint-7",
  "package-json-has-no-patch-precision",
  "package-json-has-no-fat-color-dependency-use-shuutils-or-nanocolors",
  "package-json-has-no-fat-fs-extra-dependency-use-native-fs",
  "package-json-has-no-utopian-shuunen-stack-dependency",
  "package-json-has-no-fat-slow-jsdom-dependency-use-happy-dom-instead",
  "package-json-has-no-fat-task-runner-use-npm-run-xyz-npm-run-abc-for-sequential-or-zero-deps-package-npm-parallel",
  "readme-md-has-no-link-to-deprecated-netlify-com",
  "readme-md-has-no-links-without-https-scheme",
  "readme-md-has-no-crlf-windows-carriage-return",
  "readme-md-has-no-star-flavored-list",
  "has-a-repo-checker-json-file",
]
`;

exports[`check C data/tsProject 2`] = `
[
  "has-a-editorconfig-file",
  "editorconfig-does-not-have-space-indent-indent-style-space",
  "editorconfig-does-not-have-indent-size-of-2-indent-size-2",
  "editorconfig-does-not-have-unix-style-line-endings-end-of-line-lf",
  "editorconfig-does-not-have-utf-8-encoding-charset-utf-8",
  "has-a-gitignore-file",
  "has-a-license-file",
  "license-does-not-have-a-mit-title-mit-license",
  "npmrc-file-should-be-present-when-using-pre-post-task-in-package-json-to-set-enable-pre-post-scripts-option",
  "has-a-nvmrc-file",
  "package-json-does-not-have-a-name-property-name",
  "package-json-does-not-have-a-typescript-build-or-check-btscb",
  "package-json-does-not-have-a-typescript-lint-eslint-ts",
  "has-a-readme-md-file",
  "readme-md-does-not-have-a-title-sw",
  "renovate-json-does-not-have-an-extends-section-extends",
  "renovate-json-does-not-have-a-base-config-config-base",
  "renovate-json-does-not-have-a-dashboard-setting-to-false-dependency-dashboard-false",
  "renovate-json-does-not-have-an-auto-merge-preset-automerge-all",
]
`;

exports[`check C data/tsProject 3`] = `
[
  "use-dependency-cruiser",
  "editorconfig-could-have-whitespace-trailing-trim-trailing-whitespace-true",
  "editorconfig-could-have-final-new-line-rule-insert-final-newline-true",
  "eslintrc-json-could-have-unicorn-rules-extend-plugin-unicorn-all",
  "eslintrc-json-current-eslintrc-json-has-only-2-of-the-77-custom-rules-in-repo-checker-eslintrc-json",
  "gitignore-could-have-node-modules-node-modules",
  "nvmrc-could-have-a-recent-lts-node-version-168-d-d",
  "package-json-main-file-maximum-size-is-specified-in-data-file-ex-max-size-ko-100",
  "package-json-could-have-a-schema-declaration-like-schema-https-json-schemastore-org-package-43m30m-fixable-39m49m",
  "package-json-could-have-a-bugs-property-bugs",
  "package-json-could-have-a-files-property-files-n",
  "package-json-could-have-a-homepage-property-homepage",
  "package-json-could-have-a-keywords-property-keywords-n",
  "package-json-could-have-a-private-property-private-falsetrue-n",
  "package-json-could-have-a-repository-property-repository-n",
  "package-json-could-have-a-pre-script-for-version-automation-like-preversion-npm-run-check",
  "package-json-could-have-a-post-script-for-version-automation-like-postversion-git-push-git-push-tags",
  "package-json-could-have-a-final-echo-for-task-build-build-echo-wsbuild-w",
  "package-json-one-unit-testing-dependency-from-vitest-mocha-uvu",
  "package-json-one-coverage-dependency-from-nyc-c8",
  "package-json-ts-node-without-transpile-only-detected-in-files-package-json",
  "readme-md-could-have-a-svg-logo-logofolder-logo-svg",
  "readme-md-could-have-a-demo-screen-or-gif-demofolder-demo-gif",
  "readme-md-could-have-a-project-license-badge-project-licensehttps-img-shields-io-github-license-shuunen-a-great-repo-svgcolorinformationalhttps-github-com-shuunen-a-great-repo-blob-master-license-43m30m-fixable-39m49m",
  "readme-md-could-have-a-thanks-section-thanks",
  "has-a-renovate-json-file",
  "renovate-json-could-have-a-schema-declaration-like-schema-https-docs-renovatebot-com-renovate-schema-json-43m30m-fixable-39m49m",
  "renovate-json-could-have-a-preserve-semver-ranges-preset-preserve-semver-ranges-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-schema-declaration-like-schema-https-json-schemastore-org-tsconfig-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-an-include-section-include-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-allow-unreachable-code-compiler-option-allow-unreachable-code-false-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-allow-unused-labels-compiler-option-allow-unused-labels-false-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-check-js-compiler-option-check-js-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-es-module-interop-compiler-option-es-module-interop-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-exact-optional-property-types-compiler-option-exact-optional-property-types-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-force-consistent-casing-in-file-names-compiler-option-force-consistent-casing-in-file-names-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-fallthrough-cases-in-switch-compiler-option-no-fallthrough-cases-in-switch-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-implicit-any-compiler-option-no-implicit-any-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-implicit-override-compiler-option-no-implicit-override-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-implicit-returns-compiler-option-no-implicit-returns-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-property-access-from-index-signature-compiler-option-no-property-access-from-index-signature-false-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-unchecked-indexed-access-compiler-option-no-unchecked-indexed-access-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-unused-locals-compiler-option-no-unused-locals-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-no-unused-parameters-compiler-option-no-unused-parameters-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-skip-lib-check-compiler-option-skip-lib-check-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-strict-compiler-option-strict-true-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-out-dir-compiler-option-ex-out-dir-dist-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-module-resolution-compiler-option-ex-module-resolution-node-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-target-compiler-option-ex-target-es2020-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-non-empty-lib-compiler-option-ex-lib-esnext-43m30m-fixable-39m49m",
  "tsconfig-json-could-have-a-non-empty-types-compiler-option-ex-types-node-fs-promises-43m30m-fixable-39m49m",
]
`;
